### PR TITLE
Updates cothority-ts

### PR DESCRIPTION
**Pull request by ineiti**
Depends on #1853 and #1845 - will rebase against master once those are in. For testing, please rebase on those.

Making cothority-ts more consistent within itself:
- worked on *-instance.ts to always have same signature for `constructor`, `fromByzCoin`, `spawn`, and `create`
- removed getters/setters that only pass through a private variable
- updated `spawner-instance` and `credentials-instance` to fit latest version of personhood project
- allow correct signing for multi-instruction transactions with different signers for each transaction
- removed all circular dependencies
- made logger output variable-type
- added constants for commands and arguments to instances
- added recovery mode for credentials, so that it can be recovered by a threshold of other identities
- added calypso-instance for creating write- and read-instances
- added personhood-rpc for interaction with the personhood service
- added ring signatures for use in personhood-parties

### CODE REVIEW

**Comment by Gilthoniel** [comment]
Ok so I was going to start to review this in details ... but this is simply too huge. There are plenty of backwards incompatible changes. I guess we need a chat with Jeff about how to handle that.

I went through the code and this is going to be a lot of comments, and I'm afraid this will go out of control. And I'm pretty sure you won't be happy with what I have to say so I wonder if we should just take a moment to go through this together after the dependencies are merged.

**Comment by ineiti** [comment]
OK, sounds good.

**Comment by jeffallen** [comment]
@Gilthoniel and I discussed backwards compatibility, and we agree that for immature APIs, esp client side, that accepting big refactoring right now is ok.

Compare with: https://github.com/dedis/cothority/pull/1845#discussion_r281200255

File: external/js/cothority/src/byzcoin/contracts/darc-instance.ts [hunk 1]
```diff
@@ -1,45 +1,69 @@
...
      */
-    getDarc(): Darc {
-        return this.darc;
+    static create(rpc: ByzCoinRPC,
+                  d: Darc): DarcInstance {
+        return new DarcInstance(rpc, Instance.fromFields(d.getBaseID(), DarcInstance.contractID, d.getBaseID(),
+            d.toBytes()));
+    }
+
+    darc: Darc;
```
Comment (Gilthoniel) [inline external/js/cothority/src/byzcoin/contracts/darc-instance.ts]: private _darc: Darc;

get darc() {}

File: external/js/cothority/src/byzcoin/contracts/darc-instance.ts [hunk 2]
```diff
@@ -48,57 +72,168 @@ export default class DarcInstance {
...
-        const proof = await this.rpc.getProof(this.darc.getGenesisDarcID());
-        this.darc = Darc.fromProof(this.darc.getGenesisDarcID(), proof);
+        const proof = await this.rpc.getProof(this.darc.getBaseID());
+        const inst = await proof.getVerifiedInstance(this.rpc.getGenesis().computeHash(), DarcInstance.contractID);
+        this.darc = Darc.decode(inst.data);
 
         return this;
     }
 
+    getSignerExpression(): Buffer {
```
Comment (Gilthoniel) [inline external/js/cothority/src/byzcoin/contracts/darc-instance.ts]: comment

File: external/js/cothority/src/byzcoin/contracts/darc-instance.ts [hunk 3]
```diff
@@ -48,57 +72,168 @@ export default class DarcInstance {
...
+        for (const rule of this.darc.rules.list) {
+            if (rule.action === Darc.ruleSign) {
+                return rule.expr;
+            }
+        }
+        throw new Error("This darc doesn't have a sign expression");
+    }
+
+    getSignerDarcIDs(): InstanceID[] {
```
Comment (Gilthoniel) [inline external/js/cothority/src/byzcoin/contracts/darc-instance.ts]: comment

File: external/js/cothority/src/byzcoin/contracts/darc-instance.ts [hunk 4]
```diff
@@ -48,57 +72,168 @@ export default class DarcInstance {
...
+        });
+        return ret;
+    }
+
     /**
      * Request to evolve the existing darc using the new darc and wait for
      * the block inclusion
      *
-     * @param newDarc The new darc
+     * @param newD The new darc
```
Comment (Gilthoniel) [inline external/js/cothority/src/byzcoin/contracts/darc-instance.ts]: newDarc as a static and keep the name here

File: external/js/cothority/src/byzcoin/contracts/darc-instance.ts [hunk 5]
```diff
@@ -48,57 +72,168 @@ export default class DarcInstance {
...
+    }
+}
+
+/**
+ * Create a list of rules with basic permissions for owners and signers
+ * @param owners those allow to evolve the darc
+ * @param signers those allow to sign
+ * @returns the list of rules
+ */
+export function initRules(owners: IIdentity[], signers: IIdentity[]): Rules {
```
Comment (Gilthoniel) [inline external/js/cothority/src/byzcoin/contracts/darc-instance.ts]: move them as static so we do the same thing everywhere

File: external/js/cothority/src/byzcoin/contracts/spawner-instance.ts [hunk 6]
```diff
@@ -58,84 +86,18 @@ export default class SpawnerInstance {
     /**
      * Initializes using an existing coinInstance from ByzCoin
      *
-     * @param bc
-     * @param iid
+     * @param bc an initialized byzcoin RPC instance
+     * @param iid the instance-ID of the spawn-instance
      */
     static async fromByzcoin(bc: ByzCoinRPC, iid: InstanceID): Promise<SpawnerInstance> {
```
Comment (Gilthoniel) [inline external/js/cothority/src/byzcoin/contracts/spawner-instance.ts]: add the wait/repeat parameter with default values

File: external/js/cothority/src/byzcoin/contracts/spawner-instance.ts [hunk 7]
```diff
@@ -177,60 +131,60 @@ export default class SpawnerInstance {
...
-        const d = SpawnerInstance.prepareUserDarc(pubKey, alias);
-        try {
-            const darc = await DarcInstance.fromByzcoin(this.rpc, d.getGenesisDarcID());
-            Log.warn("this darc is already registerd");
-            return darc;
-        } catch (e) {
-            // darc already exists
-        }
-
+    async spawnDarc(coin: CoinInstance, signers: Signer[], ...darcs: Darc[]): Promise<DarcInstance[]> {
```
Comment (Gilthoniel) [inline external/js/cothority/src/byzcoin/contracts/spawner-instance.ts]: spawnDarcs

Comment (Gilthoniel) [reply external/js/cothority/src/byzcoin/contracts/spawner-instance.ts]: comment

File: external/js/cothority/src/byzcoin/contracts/spawner-instance.ts [hunk 8]
```diff
@@ -177,60 +131,60 @@ export default class SpawnerInstance {
...
+        await ctx.updateCountersAndSign(this.rpc, [signers]);
 
         await this.rpc.sendTransactionAndWait(ctx);
 
-        return DarcInstance.fromByzcoin(this.rpc, d.getGenesisDarcID());
+        const dis = [];
+        for (const da of darcs) {
+            for (let i = 0; i < 10; i++) {
+                try {
+                    dis.push(await DarcInstance.fromByzcoin(this.rpc, da.getBaseID()));
```
Comment (Gilthoniel) [inline external/js/cothority/src/byzcoin/contracts/spawner-instance.ts]: wait parameter

File: external/js/cothority/src/byzcoin/contracts/spawner-instance.ts [hunk 9]
```diff
@@ -425,12 +376,52 @@ export default class SpawnerInstance {
 
         return rpsi;
     }
+
+    async spawnCalypsoWrite(coinInst: CoinInstance, signers: Signer[], lts: LongTermSecret, key: Buffer,
```
Comment (Gilthoniel) [inline external/js/cothority/src/byzcoin/contracts/spawner-instance.ts]: remember to move the tests

File: external/js/cothority/src/byzcoin/contracts/spawner-instance.ts [hunk 10]
```diff
@@ -494,55 +487,91 @@ export class SpawnerStruct extends Message<SpawnerStruct> {
...
     costParty: Long;
+
+    [k: string]: Long;
 }
 
 /**
  * Parameters to create a spawner instance
  */
 interface ICreateSpawner {
-    [k: string]: any;
```
Comment (Gilthoniel) [inline external/js/cothority/src/byzcoin/contracts/spawner-instance.ts]: revert this

File: external/js/cothority/src/byzcoin/contracts/spawner-instance.ts [hunk 11]
```diff
@@ -494,55 +487,91 @@ export class SpawnerStruct extends Message<SpawnerStruct> {
...
     signers: Signer[];
     orgs: CredentialInstance[];
     desc: PopDesc;
     reward: Long;
+
+    [k: string]: any;
+}
+
+/**
+ * Parameters to create a rock-paper-scisors game
```
Comment (Gilthoniel) [inline external/js/cothority/src/byzcoin/contracts/spawner-instance.ts]: wrong comment

File: external/js/cothority/src/byzcoin/instance.ts [hunk 12]
```diff
@@ -17,7 +18,7 @@ export default class Instance {
             throw new Error(`key not in proof: ${key.toString("hex")}`);
         }
 
-        return new Instance(key, p.contractID, p.darcID, p.value);
+        return Instance.fromFields(key, p.contractID, p.darcID, p.value);
```
Comment (Gilthoniel) [inline external/js/cothority/src/byzcoin/instance.ts]: same as above

File: external/js/cothority/src/byzcoin/instance.ts [hunk 13]
```diff
@@ -27,20 +28,52 @@ export default class Instance {
      * @returns the instance if it exists
      */
     static async fromByzCoin(rpc: ByzCoinRPC, id: InstanceID): Promise<Instance> {
-        const p = await rpc.getProof(id);
+        const p = await rpc.getProof(id, 1);
```
Comment (Gilthoniel) [inline external/js/cothority/src/byzcoin/instance.ts]: wait parameter with default values

File: external/js/cothority/src/byzcoin/instance.ts [hunk 14]
```diff
@@ -27,20 +28,52 @@ export default class Instance {
...
-    readonly darcID: InstanceID;
-    readonly data: Buffer;
+    /**
+     * Creates a new instance from separated fields.
+     * @param id
+     * @param contractID
+     * @param darcID
+     * @param data
+     */
+    static fromFields(id: InstanceID, contractID: string, darcID: InstanceID, data: Buffer): Instance {
```
Comment (Gilthoniel) [inline external/js/cothority/src/byzcoin/instance.ts]: you can remove that

File: external/js/cothority/src/byzcoin/instance.ts [hunk 15]
```diff
@@ -27,20 +28,52 @@ export default class Instance {
...
+        const obj = JSON.parse(buf.toString());
+        return new Instance({
+            contractID: obj.contractID,
+            darcID: Buffer.from(obj.darcID),
+            data: Buffer.from(obj.data),
+            id: Buffer.from(obj.id),
+        });
+    }
+
+    id: InstanceID;
```
Comment (Gilthoniel) [inline external/js/cothority/src/byzcoin/instance.ts]: readonly

File: external/js/cothority/src/byzcoin/proof.ts [hunk 16]
```diff
@@ -88,6 +72,38 @@ export default class Proof extends Message<Proof> {
...
+        if (err != null) {
+            return Promise.reject(err);
+        }
+        if (!this.exists(this.key)) {
+            return Promise.reject("cannot return an Instance from a proof of absence");
+        }
+        if (this.contractID !== contractID) {
+            return Promise.reject("not of correct contractID");
+        }
+        return Instance.fromFields(this.key, this.contractID, this.darcID, this.value);
```
Comment (Gilthoniel) [inline external/js/cothority/src/byzcoin/proof.ts]: new Instance()

File: external/js/cothority/src/byzcoin/proof.ts [hunk 17]
```diff
@@ -105,7 +121,12 @@ export default class Proof extends Message<Proof> {
             return new Error("invalid root");
         }
 
-        let publics = this.latest.roster.getServicePublics(SkipchainRPC.serviceName);
+        if (SkipchainRPC.disableSignatureVerification) {
```
Comment (Gilthoniel) [inline external/js/cothority/src/byzcoin/proof.ts]: remove that

File: external/js/cothority/src/byzcoin/proof.ts [hunk 18]
```diff
@@ -1,10 +1,12 @@
 import { createHash } from "crypto";
-import _ from "lodash";
+import * as _ from "lodash";
```
Comment (Gilthoniel) [inline external/js/cothority/src/byzcoin/proof.ts]: import * as _ is really bad

File: external/js/cothority/src/byzcoin/proto/requests.ts [hunk 19]
```diff
@@ -1,3 +1,4 @@
+import Long from "long";
```
Comment (Gilthoniel) [inline external/js/cothority/src/byzcoin/proto/requests.ts]: you can remove it

File: external/js/cothority/src/calypso/keccak.d.ts [hunk 20]
```diff
@@ -0,0 +1 @@
+declare module "keccak";
```
Comment (Gilthoniel) [inline external/js/cothority/src/calypso/keccak.d.ts]: this file should be removed

File: external/js/cothority/src/log.ts [hunk 21]
```diff
@@ -1,128 +1,181 @@
...
+
 const defaultLvl = 2;
-const lvlStr = ["E", "W", "I", "!4", "!3", "!2", "!1", "P", "1", "2", "3", "4"];
 
-const regex = /([^/]+:[0-9]+):[0-9]+/;
+const lvlStr = ["E ", "W ", "I ", "!4", "!3", "!2", "!1", "P ", " 1", " 2", " 3", " 4"];
+
+export class LogC {
 
-export class Logger {
```
Comment (Gilthoniel) [inline external/js/cothority/src/log.ts]: revert

File: external/js/cothority/src/network/proto.ts [hunk 22]
```diff
@@ -2,7 +2,7 @@ import { Point, PointFactory } from "@dedis/kyber";
 import { createHash } from "crypto";
 import { Message, Properties } from "protobufjs/light";
 import UUID from "pure-uuid";
-import toml from "toml";
+import * as toml from "toml";
```
Comment (Gilthoniel) [inline external/js/cothority/src/network/proto.ts]: same story

File: external/js/cothority/src/status/status-rpc.ts [hunk 23]
```diff
@@ -37,6 +37,6 @@ export default class StatusRPC {
 
         this.conn[index].setTimeout(this.timeout);
 
-        return this.conn[index].send(new StatusRequest(), StatusResponse);
+        return (await this.conn[index].send(new StatusRequest(), StatusResponse)) as StatusResponse;
```
Comment (Gilthoniel) [inline external/js/cothority/src/status/status-rpc.ts]: it's returning a promise

File: external/js/cothority/src/byzcoin/proto/requests.ts (hunk 19 above)
Comment (ineiti) [reply external/js/cothority/src/byzcoin/proto/requests.ts]: I get the following tslint error if I remove the import:
```
        const signer = SignerEd25519.fromBytes(ephemeral);
```
But I'm not sure if my tslint is badly configured.

File: external/js/cothority/src/calypso/keccak.d.ts (hunk 20 above)
Comment (ineiti) [reply external/js/cothority/src/calypso/keccak.d.ts]: if I run `npm run test` without that file, I get
```
TSError: ⨯ Unable to compile TypeScript:
src/darc/darc.ts(67,22): error TS2686: 'Long' refers to a UMD global, but the current file is a module. Consider adding an import instead.
```

File: external/js/cothority/src/byzcoin/contracts/spawner-instance.ts (hunk 6 above)
Comment (ineiti) [reply external/js/cothority/src/byzcoin/contracts/spawner-instance.ts]: Added it to all the `Instance`s.

**Pull request merged**